### DISCUSSION TITLE
Need sudo -v before when it's scripted

### DISCUSSION THREAD

**Discussion by kud** [Homebrew/discussions#3803] [Getting started]
Hey,

I've scripted the installation and for a reason I don't understand yet, I need first to `sudo -v` to be sure that my current account is sudo.

Any idea why?

```
if ! type "brew" > /dev/null; then
  echo_space
  echo_title_install "homebrew"
  sudo -v
  /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
fi
```

I don't need it outside my script if I run the command `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` directly.

Thanks for your help,

**Comment by SMillerDev** [comment]
The normal install also requires root access.

**Reply by kud** [reply]
I'm sorry. It's not my question. Apparently Homebrew checks first if it needs sudo access and seems to ask to be in sudo.

It's the case when I start the script itself but it does not happen when I start it via a script.

**Comment by gromgit** [comment]
> I don't need it outside my script if I run the command `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` directly.

`sudo` caches credentials for a period (15 mins by default, I think), so if you'd run a `sudo` command recently (or even `sudo -v`, which extends the cache timeout for another 15 mins), that's why.

**Reply by kud** [reply]
I see. Thank you.

**Reply by gromgit** [reply]
To be clear, the official installer [absolutely uses `sudo`](https://github.com/Homebrew/install/blob/77146edc9f3757a8e90f10feff4119486484dc52/install.sh#L239-L253), so running `sudo -v` in your script is useless. Either:
1. you have no `sudo` credentials cached, in which case your invocation prompts you for a password (so the installer's `sudo` invocations won't), or 
2. your `sudo` credentials are still cached, then _none_ of the `sudo` invocations will prompt for a password.

**Reply by kud** [reply]
The thing is if I do not do `sudo -v`, Homebrew checks the rights and just after says `This user needs to be sudo to continue!`.

`sudo -v` was my only way to make the checking work correctly on my computer.

**Reply by gromgit** [reply]
In that case, I'm guessing that one or more of the following is true, in decreasing order of probability:
1. you're redirecting stdin of your script from `/dev/null` or some other file: https://github.com/Homebrew/install/blob/77146edc9f3757a8e90f10feff4119486484dc52/install.sh#L89-L94
2. you're setting the `NONINTERACTIVE` environment variable: https://github.com/Homebrew/install/blob/77146edc9f3757a8e90f10feff4119486484dc52/install.sh#L83
3. you're setting the `CI` environment variable: https://github.com/Homebrew/install/blob/77146edc9f3757a8e90f10feff4119486484dc52/install.sh#L85-L88

In all the above, the installer will invoke `sudo` with the `-n` flag, which tells `sudo` to never prompt for a password, which would of course fail spectacularly if you didn't already have `sudo` credentials cached. That's why you had to run `sudo -v` separately.

**Reply by kud** [reply]
Thank you. I feel you're right about the 2/.

If you're interested, all is here: https://github.com/kud/my
